File: TextCascade.Server/Protocol.cs
using System.Buffers.Text;
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.Json.Serialization.Metadata;

namespace TextCascade.Server;

public enum MessageKind
{
    Unknown,
    Hello,
    Clip,
    Pong,
}

public enum ProtocolErrorCode
{
    InvalidMessage,
    TextTooLarge,
    FrameTooLarge,
    EmptyText,
    RateLimited,
    HelloTimeout,
    ServerBusy,
}

public sealed record ProtocolError(ProtocolErrorCode Code, string Message, string? ReferenceId = null)
{
    public string CodeName => Code switch
    {
        ProtocolErrorCode.TextTooLarge => "text_too_large",
        ProtocolErrorCode.FrameTooLarge => "frame_too_large",
        ProtocolErrorCode.EmptyText => "empty_text",
        ProtocolErrorCode.RateLimited => "rate_limited",
        ProtocolErrorCode.HelloTimeout => "hello_timeout",
        ProtocolErrorCode.ServerBusy => "server_busy",
        _ => "invalid_message",
    };
}

public readonly record struct Result<T>
{
    public Result(T? value, ProtocolError? error = null)
    {
        Value = value;
        Error = error;
    }

    public T? Value { get; }

    public ProtocolError? Error { get; }

    public bool IsSuccess => Error is null;

    public static Result<T> Success(T value) => new(value, null);

    public static Result<T> Failure(ProtocolErrorCode code, string message, string? referenceId = null) =>
        new(default, new ProtocolError(code, message, referenceId));
}

public sealed record ClipSnapshot(
    string Payload,
    bool Encrypted,
    string Hash,
    DateTimeOffset LocalModifiedAtUtc);

public sealed record ClientHello(
    string ClientId,
    string ClientName,
    ulong LastServerVersion,
    ClipSnapshot? Snapshot);

public sealed record ClientClip(
    string Id,
    string Payload,
    bool Encrypted,
    string Hash);

public sealed record ClientPong(DateTimeOffset ClientTimeUtc);

public sealed record ClientMessage(MessageKind Kind, object Message);

public sealed record LatestText(
    [property: JsonPropertyName("payload")] string Payload,
    [property: JsonPropertyName("version")] ulong Version,
    [property: JsonPropertyName("hash")] string Hash,
    [property: JsonPropertyName("encrypted")] bool Encrypted,
    [property: JsonPropertyName("fromClientId")] string FromClientId,
    [property: JsonPropertyName("fromClientName")] string FromClientName,
    [property: JsonPropertyName("updatedAtUtc")] DateTimeOffset UpdatedAtUtc)
{
    public static LatestText From(ClipSnapshot snapshot, ulong version, string clientId, string clientName)
    {
        return new LatestText(
            snapshot.Payload,
            version,
            snapshot.Hash,
            snapshot.Encrypted,
            clientId,
            clientName,
            snapshot.LocalModifiedAtUtc);
    }
}

[JsonSerializable(typeof(WelcomeMessage))]
[JsonSerializable(typeof(ClipMessage))]
[JsonSerializable(typeof(ClipAckMessage))]
[JsonSerializable(typeof(PingMessage))]
[JsonSerializable(typeof(ByeMessage))]
[JsonSerializable(typeof(ProtocolErrorMessage))]
[JsonSourceGenerationOptions(DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull)]
internal sealed partial class ServerJsonContext : JsonSerializerContext
{
    public static ServerJsonContext Configured { get; }

    public static JsonSerializerOptions SerializationOptions { get; } = new(JsonSerializerDefaults.Web)
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        WriteIndented = false,
        Converters = { new UtcSecondDateTimeConverter() },
    };

    static ServerJsonContext() => Configured = new ServerJsonContext(SerializationOptions);
}

public sealed record WelcomeMessage(
    [property: JsonPropertyName("type")] string Type,
    [property: JsonPropertyName("protocolVersion")] int ProtocolVersion,
    [property: JsonPropertyName("latest")] LatestText? Latest);

public sealed record ClipMessage(
    [property: JsonPropertyName("type")] string Type,
    [property: JsonPropertyName("version")] ulong Version,
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("payload")] string Payload,
    [property: JsonPropertyName("encrypted")] bool Encrypted,
    [property: JsonPropertyName("hash")] string Hash,
    [property: JsonPropertyName("fromClientId")] string FromClientId,
    [property: JsonPropertyName("fromClientName")] string FromClientName,
    [property: JsonPropertyName("updatedAtUtc")] DateTimeOffset UpdatedAtUtc);

public sealed record ClipAckMessage(
    [property: JsonPropertyName("type")] string Type,
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("version")] ulong Version,
    [property: JsonPropertyName("updatedAtUtc")] DateTimeOffset UpdatedAtUtc);

public sealed record PingMessage(
    [property: JsonPropertyName("type")] string Type,
    [property: JsonPropertyName("serverTimeUtc")] DateTimeOffset ServerTimeUtc);

public sealed record ByeMessage(
    [property: JsonPropertyName("type")] string Type,
    [property: JsonPropertyName("reason")] string Reason);

public sealed record ProtocolErrorMessage(
    [property: JsonPropertyName("type")] string Type,
    [property: JsonPropertyName("code")] string Code,
    [property: JsonPropertyName("message")] string Message,
    [property: JsonPropertyName("referenceId")] string? ReferenceId);

internal sealed class UtcSecondDateTimeConverter : JsonConverter<DateTimeOffset>
{
    public override DateTimeOffset Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        => reader.TokenType == JsonTokenType.String && DateTimeOffset.TryParse(reader.GetString(), CultureInfo.InvariantCulture, DateTimeStyles.None, out var value)
            ? value.ToUniversalTime()
            : throw new JsonException("Invalid date/time value.");

    public override void Write(Utf8JsonWriter writer, DateTimeOffset value, JsonSerializerOptions options)
        => writer.WriteStringValue(value.ToUniversalTime().ToString("yyyy-MM-dd'T'HH:mm:ss'Z'", CultureInfo.InvariantCulture));
}

public static class Protocol
{
    public const int ProtocolVersion = 1;

    public const int MaxNameBytes = 128;

    public const int MaxIdBytes = 128;

    public const int MaxHashBytes = 4096;

    public static byte[] SerializeMessage<T>(T message, JsonTypeInfo<T> typeInfo) =>
        JsonSerializer.SerializeToUtf8Bytes(message, typeInfo);

    public static byte[] SerializeWelcome(LatestText? latest, LimitsConfig _)
    {
        var message = new WelcomeMessage("welcome", ProtocolVersion, latest);
        return SerializeMessage(message, ServerJsonContext.Configured.WelcomeMessage);
    }

    public static byte[] SerializeClip(string id, LatestText latest)
    {
        var message = new ClipMessage(
            "clip",
            latest.Version,
            id,
            latest.Payload,
            latest.Encrypted,
            latest.Hash,
            latest.FromClientId,
            latest.FromClientName,
            latest.UpdatedAtUtc);
        return SerializeMessage(message, ServerJsonContext.Configured.ClipMessage);
    }

    public static byte[] SerializeClipAck(string id, LatestText latest)
    {
        var message = new ClipAckMessage("clip_ack", id, latest.Version, latest.UpdatedAtUtc);
        return SerializeMessage(message, ServerJsonContext.Configured.ClipAckMessage);
    }

    public static byte[] SerializePing(DateTimeOffset nowUtc) =>
        SerializeMessage(new PingMessage("ping", nowUtc), ServerJsonContext.Configured.PingMessage);

    public static byte[] SerializeBye(string reason = "server_shutdown") =>
        SerializeMessage(new ByeMessage("bye", reason), ServerJsonContext.Configured.ByeMessage);

    public static byte[] SerializeProtocolError(ProtocolError error) =>
        SerializeMessage(
            new ProtocolErrorMessage("error", error.CodeName, error.Message, error.ReferenceId),
            ServerJsonContext.Configured.ProtocolErrorMessage);

    public static byte[] SerializeLoginResponse(AuthToken token, RuntimeConfig config, bool needsRehash = false)
    {
        using var stream = new MemoryStream();
        using var writer = new Utf8JsonWriter(stream);
        writer.WriteStartObject();
        writer.WriteString("token", token.CompactToken);
        writer.WriteString("expiresAtUtc", DateTimeOffset.FromUnixTimeSeconds(token.Payload.ExpiresAtUnix).ToUniversalTime().ToString("O"));
        writer.WriteNumber("protocolVersion", ProtocolVersion);
        writer.WriteNumber("maxTextBytes", config.Limits.MaxTextBytes);
        writer.WriteNumber("helloTimeoutSeconds", config.Limits.HelloTimeoutSeconds);
        writer.WriteNumber("heartbeatIntervalSeconds", config.Limits.HeartbeatIntervalSeconds);
        writer.WriteNumber("heartbeatTimeoutSeconds", config.Limits.HeartbeatTimeoutSeconds);
        if (needsRehash)
        {
            writer.WriteBoolean("needsRehash", true);
        }
        writer.WriteEndObject();
        writer.Flush();
        return stream.ToArray();
    }

    public static ParseResult ParseClientMessage(ReadOnlySpan<byte> frame, RuntimeConfig config)
    {
        using var document = TryParseJson(frame, out var parseError);
        if (parseError is not null)
        {
            return ParseResult.Failure(parseError);
        }

        var root = document!.RootElement;
        if (root.ValueKind != JsonValueKind.Object)
        {
            return ParseResult.Failure(Invalid("Root must be an object."));
        }

        if (!TryGetString(root, "type", null, out var type))
        {
            return ParseResult.Failure(Invalid("Missing or invalid type.", GetReferenceId(root)));
        }

        if (!HasUniqueKnownProperties(root, type switch
        {
            "hello" => new[] { "type", "clientId", "clientName", "lastServerVersion", "snapshot" },
            "clip" => new[] { "type", "id", "payload", "encrypted", "hash" },
            "pong" => new[] { "type", "clientTimeUtc" },
            _ => Array.Empty<string>(),
        }, out var propertyName))
        {
            return ParseResult.Failure(Invalid($"Unknown or duplicate field: {propertyName}.", GetReferenceId(root)));
        }

        return type switch
        {
            "hello" => ParseHello(root, config),
            "clip" => ParseClip(root, config),
            "pong" => ParsePong(root),
            _ => ParseResult.Failure(Invalid("Unknown message type.", GetReferenceId(root))),
        };
    }

    private static ParseResult ParseHello(JsonElement root, RuntimeConfig config)
    {
        if (!TryGetString(root, "clientId", MaxNameBytes, out var clientId) || clientId.Length == 0)
        {
            return ParseResult.Failure(Invalid("clientId must be 1-128 bytes."));
        }

        TryGetString(root, "clientName", MaxNameBytes, out var clientName);

        if (!TryGetUInt64(root, "lastServerVersion", out var lastVersion))
        {
            return ParseResult.Failure(Invalid("lastServerVersion must be a non-negative integer."));
        }

        if (root.TryGetProperty("snapshot", out var snapshotElement))
        {
            if (snapshotElement.ValueKind != JsonValueKind.Object && snapshotElement.ValueKind != JsonValueKind.Null)
            {
                return ParseResult.Failure(Invalid("snapshot must be an object or null."));
            }

            if (snapshotElement.ValueKind == JsonValueKind.Object)
            {
                if (!HasUniqueKnownProperties(snapshotElement,
                    new[] { "payload", "encrypted", "hash", "localModifiedAtUtc" },
                    out var propertyName))
                {
                    return ParseResult.Failure(Invalid($"Unknown or duplicate field: {propertyName}."));
                }

                if (!TryGetSnapshot(snapshotElement, config, out var snapshot, out var snapshotError))
                {
                    return ParseResult.Failure(snapshotError!);
                }

                var hello = new ClientHello(clientId, clientName, lastVersion, snapshot);
                return ValidateHello(hello, config) ? ParseResult.Success(MessageKind.Hello, hello)
                    : ParseResult.Failure(Invalid("hello validation failed."));
            }
        }

        var plainHello = new ClientHello(clientId, clientName, lastVersion, null);
        return ValidateHello(plainHello, config) ? ParseResult.Success(MessageKind.Hello, plainHello)
            : ParseResult.Failure(Invalid("hello validation failed."));
    }

    private static bool TryGetSnapshot(JsonElement element, RuntimeConfig config, out ClipSnapshot? snapshot, out ProtocolError? error)
    {
        snapshot = null;
        if (!TryGetString(element, "payload", null, out var payload) || string.IsNullOrEmpty(payload))
        {
            error = Invalid("snapshot.payload is required and must not be empty.");
            return false;
        }

        if (!ValidatePayloadSize(payload, config, out var payloadError))
        {
            error = payloadError;
            return false;
        }

        if (!TryGetBoolean(element, "encrypted", out var encrypted))
        {
            error = Invalid("snapshot.encrypted must be a boolean.");
            return false;
        }

        if (!TryGetString(element, "hash", MaxHashBytes, out var hash) || hash.Length == 0)
        {
            error = Invalid("snapshot.hash is required.");
            return false;
        }

        if (!TryGetUtcDateTime(element, "localModifiedAtUtc", out var modified))
        {
            error = Invalid("snapshot.localModifiedAtUtc must be UTC RFC3339.");
            return false;
        }

        snapshot = new ClipSnapshot(payload, encrypted, hash, modified);
        error = null;
        return true;
    }

    private static ParseResult ParseClip(JsonElement root, RuntimeConfig config)
    {
        if (!TryGetString(root, "id", MaxIdBytes, out var id) || id.Length == 0)
        {
            return ParseResult.Failure(Invalid("id must be 1-128 bytes."));
        }

        if (!TryGetString(root, "payload", null, out var payload) || string.IsNullOrEmpty(payload))
        {
            return ParseResult.Failure(new ProtocolError(ProtocolErrorCode.EmptyText, "payload must not be empty.", id));
        }

        if (!ValidatePayloadSize(payload, config, out var error))
        {
            return ParseResult.Failure(new ProtocolError(error!.Code, error.Message, id));
        }

        if (!TryGetBoolean(root, "encrypted", out var encrypted))
        {
            return ParseResult.Failure(Invalid("encrypted must be a boolean.", id));
        }

        if (!TryGetString(root, "hash", MaxHashBytes, out var hash) || hash.Length == 0)
        {
            return ParseResult.Failure(Invalid("hash is required.", id));
        }

        var clip = new ClientClip(id, payload, encrypted, hash);
        return ValidateClipMessage(clip, config)
            ? ParseResult.Success(MessageKind.Clip, clip)
            : ParseResult.Failure(Invalid("clip validation failed.", id));
    }

    private static ParseResult ParsePong(JsonElement root)
    {
        if (!TryGetUtcDateTime(root, "clientTimeUtc", out var clientTime))
        {
            return ParseResult.Failure(Invalid("clientTimeUtc must be UTC RFC3339."));
        }

        var pong = new ClientPong(clientTime);
        return ParseResult.Success(MessageKind.Pong, pong);
    }

    public static bool ValidateHello(ClientHello hello, RuntimeConfig config)
    {
        return Encoding.UTF8.GetByteCount(hello.ClientId) is > 0 and <= MaxNameBytes
            && Encoding.UTF8.GetByteCount(hello.ClientName) <= MaxNameBytes
            && (hello.Snapshot is null || ValidateClipSnapshot(hello.Snapshot, config));
    }

    private static bool ValidateClipSnapshot(ClipSnapshot snapshot, RuntimeConfig config)
    {
        return snapshot.Payload.Length > 0
            && Encoding.UTF8.GetByteCount(snapshot.Payload) <= config.Limits.MaxTextBytes
            && Encoding.UTF8.GetByteCount(snapshot.Hash) <= MaxHashBytes
            && snapshot.LocalModifiedAtUtc.Offset == TimeSpan.Zero;
    }

    public static bool ValidateClipMessage(ClientClip message, RuntimeConfig config)
    {
        return message.Id.Length > 0
            && Encoding.UTF8.GetByteCount(message.Id) <= MaxIdBytes
            && message.Payload.Length > 0
            && Encoding.UTF8.GetByteCount(message.Payload) <= config.Limits.MaxTextBytes
            && message.Hash.Length > 0
            && Encoding.UTF8.GetByteCount(message.Hash) <= MaxHashBytes;
    }

    public static bool CheckFrameSize(int frameLength, RuntimeConfig config)
    {
        return frameLength is > 0 && frameLength <= config.Limits.MaxFrameBytes;
    }

    public static bool CheckPayloadSize(string payload, RuntimeConfig config)
    {
        return Encoding.UTF8.GetByteCount(payload) <= config.Limits.MaxTextBytes;
    }

    private static bool ValidatePayloadSize(string payload, RuntimeConfig config, out ProtocolError? error)
    {
        if (Encoding.UTF8.GetByteCount(payload) > config.Limits.MaxTextBytes)
        {
            error = new ProtocolError(ProtocolErrorCode.TextTooLarge, "Text exceeds maxTextBytes.");
            return false;
        }

        error = null;
        return true;
    }

    private static JsonDocument? TryParseJson(ReadOnlySpan<byte> frame, out ProtocolError? error)
    {
        try
        {
            var document = JsonDocument.Parse(frame.ToArray(), new JsonDocumentOptions
            {
                AllowTrailingCommas = false,
                CommentHandling = JsonCommentHandling.Disallow,
                MaxDepth = 3,
            });
            error = null;
            return document;
        }
        catch (JsonException exception)
        {
            error = Invalid($"Invalid JSON: {exception.Message}");
            return null;
        }
    }

    private static string? GetReferenceId(JsonElement root)
    {
        return root.ValueKind == JsonValueKind.Object
            && root.TryGetProperty("id", out var id)
            && id.ValueKind == JsonValueKind.String
            ? id.GetString()
            : null;
    }

    private static bool HasUniqueKnownProperties(JsonElement root, string[] known, out string? invalidName)
    {
        var seen = new HashSet<string>(StringComparer.Ordinal);
        foreach (var property in root.EnumerateObject())
        {
            if (!known.Contains(property.Name, StringComparer.Ordinal) || !seen.Add(property.Name))
            {
                invalidName = property.Name;
                return false;
            }
        }

        invalidName = null;
        return true;
    }

    private static bool TryGetString(JsonElement root, string name, int? maxBytes, out string value)
    {
        if (!root.TryGetProperty(name, out var element) || element.ValueKind != JsonValueKind.String)
        {
            value = string.Empty;
            return false;
        }

        value = element.GetString()!;
        return maxBytes is null || Encoding.UTF8.GetByteCount(value) <= maxBytes.Value;
    }

    private static bool TryGetBoolean(JsonElement root, string name, out bool value)
    {
        if (!root.TryGetProperty(name, out var element) || element.ValueKind != JsonValueKind.True && element.ValueKind != JsonValueKind.False)
        {
            value = false;
            return false;
        }

        value = element.GetBoolean();
        return true;
    }

    private static bool TryGetUInt64(JsonElement root, string name, out ulong value)
    {
        if (!root.TryGetProperty(name, out var element) || element.ValueKind != JsonValueKind.Number)
        {
            value = 0;
            return false;
        }

        var raw = element.GetRawText().AsSpan();
        if (raw.StartsWith('-') || raw.Contains('.') || raw.Contains('e') || raw.Contains('E') || !ulong.TryParse(raw, out value))
        {
            value = 0;
            return false;
        }

        return true;
    }

    private static bool TryGetUtcDateTime(JsonElement root, string name, out DateTimeOffset value)
    {
        if (!root.TryGetProperty(name, out var element) || element.ValueKind != JsonValueKind.String)
        {
            value = default;
            return false;
        }

        var text = element.GetString()!;
        if (!text.EndsWith('Z'))
        {
            value = default;
            return false;
        }

        if (DateTimeOffset.TryParseExact(text, "O", CultureInfoInvariant, DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal, out value)
            || DateTimeOffset.TryParseExact(text, "yyyy-MM-ddTHH:mm:ssZ", CultureInfoInvariant, DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal, out value))
        {
            return true;
        }

        value = default;
        return false;
    }

    private static ProtocolError Invalid(string message, string? referenceId = null) =>
        new(ProtocolErrorCode.InvalidMessage, message, referenceId);

    private static CultureInfo CultureInfoInvariant => CultureInfo.InvariantCulture;
}

public readonly record struct ParseResult
{
    private ParseResult(MessageKind kind, object? message, ProtocolError? error)
    {
        Kind = kind;
        Message = message;
        Error = error;
    }

    public MessageKind Kind { get; }

    public object? Message { get; }

    public ProtocolError? Error { get; }

    public bool IsSuccess => Error is null;

    public static ParseResult Success(MessageKind kind, object message) => new(kind, message, null);

    public static ParseResult Failure(ProtocolError error) => new(MessageKind.Unknown, null, error);
}

File: internal/protocol/protocol.go
// Package protocol：Protocol.cs 的结构、入站解析与校验半边（本文件），
// 出站 marshal 见 serialize.go。
package protocol

import (
	"strconv"
	"strings"
	"time"

	"github.com/long45343/TextCascade-Server/internal/config"
)

const (
	ProtocolVersion = 1

	MaxNameBytes = 128

	MaxIdBytes = 128

	MaxHashBytes = 4096
)

// ErrKind 对应 C# ProtocolErrorCode。
type ErrKind int

const (
	ErrInvalidMessage ErrKind = iota
	ErrTextTooLarge
	ErrFrameTooLarge
	ErrEmptyText
	ErrRateLimited
	ErrHelloTimeout
	ErrServerBusy
)

// Error 对应 C# ProtocolError record。
type Error struct {
	Code        ErrKind
	Message     string
	ReferenceID *string
}

func (e *Error) CodeName() string {
	switch e.Code {
	case ErrTextTooLarge:
		return "text_too_large"
	case ErrFrameTooLarge:
		return "frame_too_large"
	case ErrEmptyText:
		return "empty_text"
	case ErrRateLimited:
		return "rate_limited"
	case ErrHelloTimeout:
		return "hello_timeout"
	case ErrServerBusy:
		return "server_busy"
	default:
		return "invalid_message"
	}
}

func invalid(message string, referenceID *string) *Error {
	return &Error{Code: ErrInvalidMessage, Message: message, ReferenceID: referenceID}
}

// ClipSnapshot 对应 C# ClipSnapshot record。
type ClipSnapshot struct {
	Payload            string
	Encrypted          bool
	Hash               string
	LocalModifiedAtUtc time.Time
}

// ClientHello 对应 C# ClientHello record。
type ClientHello struct {
	ClientID          string
	ClientName        string
	LastServerVersion uint64
	Snapshot          *ClipSnapshot
}

// ClientClip 对应 C# ClientClip record。
type ClientClip struct {
	ID        string
	Payload   string
	Encrypted bool
	Hash      string
}

// ClientPong 对应 C# ClientPong record。
type ClientPong struct {
	ClientTimeUtc time.Time
}

// Kind 对应 C# MessageKind。
type Kind int

const (
	KindUnknown Kind = iota
	KindHello
	KindClip
	KindPong
)

// Message 是单一 struct（Go 无判别联合，用固定字段）。
type Message struct {
	Kind  Kind
	Hello *ClientHello
	Clip  *ClientClip
	Pong  *ClientPong
}

// LatestText 对应 C# LatestText record。
type LatestText struct {
	Payload        string
	Version        uint64
	Hash           string
	Encrypted      bool
	FromClientID   string
	FromClientName string
	UpdatedAtUtc   time.Time
}

// LatestFromSnapshot 对应 C# LatestText.From。
func LatestFromSnapshot(s *ClipSnapshot, version uint64, clientID, clientName string) *LatestText {
	return &LatestText{
		Payload:        s.Payload,
		Version:        version,
		Hash:           s.Hash,
		Encrypted:      s.Encrypted,
		FromClientID:   clientID,
		FromClientName: clientName,
		UpdatedAtUtc:   s.LocalModifiedAtUtc,
	}
}

// ParseClientMessage 顺序：jsonscan 预扫描 → 根必须 object → type 字符串 →
// 未知/重复字段检查（known 集合按类型）→ parseHello/parseClip/parsePong → 未知 type。
func ParseClientMessage(frame []byte, cfg *config.RuntimeConfig) (Message, *Error) {
	root, err := Decode(frame, 3)
	if err != nil {
		return Message{}, invalid("Invalid JSON: "+err.Error(), nil)
	}

	if !root.IsObject() {
		return Message{}, invalid("Root must be an object.", nil)
	}

	typeValue := root.Get("type")
	if typeValue == nil || !typeValue.IsString() {
		return Message{}, invalid("Missing or invalid type.", getReferenceID(root))
	}
	messageType := typeValue.Str()

	var known []string
	switch messageType {
	case "hello":
		known = []string{"type", "clientId", "clientName", "lastServerVersion", "snapshot"}
	case "clip":
		known = []string{"type", "id", "payload", "encrypted", "hash"}
	case "pong":
		known = []string{"type", "clientTimeUtc"}
	default:
		known = nil
	}

	if propertyName, ok := hasUniqueKnownProperties(root, known); !ok {
		return Message{}, invalid("Unknown or duplicate field: "+propertyName+".", getReferenceID(root))
	}

	switch messageType {
	case "hello":
		return parseHello(root, cfg)
	case "clip":
		return parseClip(root, cfg)
	case "pong":
		return parsePong(root)
	default:
		return Message{}, invalid("Unknown message type.", getReferenceID(root))
	}
}

func parseHello(root *Node, cfg *config.RuntimeConfig) (Message, *Error) {
	clientIDValue := root.Get("clientId")
	if clientIDValue == nil || !clientIDValue.IsString() || utf8Len(clientIDValue.Str()) > MaxNameBytes || clientIDValue.Str() == "" {
		return Message{}, invalid("clientId must be 1-128 bytes.", nil)
	}
	clientID := clientIDValue.Str()

	var clientName string
	if nameValue := root.Get("clientName"); nameValue != nil && nameValue.IsString() && utf8Len(nameValue.Str()) <= MaxNameBytes {
		clientName = nameValue.Str()
	}

	lastVersion, ok := tryGetUint64(root, "lastServerVersion")
	if !ok {
		return Message{}, invalid("lastServerVersion must be a non-negative integer.", nil)
	}

	if snapshotElement := root.Get("snapshot"); snapshotElement != nil {
		if !snapshotElement.IsObject() && !snapshotElement.IsNull() {
			return Message{}, invalid("snapshot must be an object or null.", nil)
		}

		if snapshotElement.IsObject() {
			if propertyName, ok := hasUniqueKnownProperties(snapshotElement, []string{"payload", "encrypted", "hash", "localModifiedAtUtc"}); !ok {
				return Message{}, invalid("Unknown or duplicate field: "+propertyName+".", nil)
			}

			snapshot, snapErr := tryGetSnapshot(snapshotElement, cfg)
			if snapErr != nil {
				return Message{}, snapErr
			}

			hello := &ClientHello{ClientID: clientID, ClientName: clientName, LastServerVersion: lastVersion, Snapshot: snapshot}
			if ValidateHello(hello, cfg) {
				return Message{Kind: KindHello, Hello: hello}, nil
			}
			return Message{}, invalid("hello validation failed.", nil)
		}
	}

	plainHello := &ClientHello{ClientID: clientID, ClientName: clientName, LastServerVersion: lastVersion, Snapshot: nil}
	if ValidateHello(plainHello, cfg) {
		return Message{Kind: KindHello, Hello: plainHello}, nil
	}
	return Message{}, invalid("hello validation failed.", nil)
}

func tryGetSnapshot(element *Node, cfg *config.RuntimeConfig) (*ClipSnapshot, *Error) {
	payloadValue := element.Get("payload")
	if payloadValue == nil || !payloadValue.IsString() || payloadValue.Str() == "" {
		return nil, invalid("snapshot.payload is required and must not be empty.", nil)
	}
	payload := payloadValue.Str()

	if utf8Len(payload) > cfg.Limits.MaxTextBytes {
		return nil, &Error{Code: ErrTextTooLarge, Message: "Text exceeds maxTextBytes."}
	}

	encryptedValue := element.Get("encrypted")
	if encryptedValue == nil || !encryptedValue.IsBool() {
		return nil, invalid("snapshot.encrypted must be a boolean.", nil)
	}
	encrypted := encryptedValue.Bool()

	hashValue := element.Get("hash")
	if hashValue == nil || !hashValue.IsString() || utf8Len(hashValue.Str()) > MaxHashBytes || hashValue.Str() == "" {
		return nil, invalid("snapshot.hash is required.", nil)
	}
	hash := hashValue.Str()

	modified, ok := ParseFlexibleTime(element, "localModifiedAtUtc")
	if !ok {
		return nil, invalid("snapshot.localModifiedAtUtc must be UTC RFC3339.", nil)
	}

	return &ClipSnapshot{Payload: payload, Encrypted: encrypted, Hash: hash, LocalModifiedAtUtc: modified}, nil
}

func parseClip(root *Node, cfg *config.RuntimeConfig) (Message, *Error) {
	idValue := root.Get("id")
	if idValue == nil || !idValue.IsString() || utf8Len(idValue.Str()) > MaxIdBytes || idValue.Str() == "" {
		return Message{}, invalid("id must be 1-128 bytes.", nil)
	}
	id := idValue.Str()

	payloadValue := root.Get("payload")
	if payloadValue == nil || !payloadValue.IsString() || payloadValue.Str() == "" {
		return Message{}, &Error{Code: ErrEmptyText, Message: "payload must not be empty.", ReferenceID: &id}
	}
	payload := payloadValue.Str()

	if utf8Len(payload) > cfg.Limits.MaxTextBytes {
		return Message{}, &Error{Code: ErrTextTooLarge, Message: "Text exceeds maxTextBytes.", ReferenceID: &id}
	}

	encryptedValue := root.Get("encrypted")
	if encryptedValue == nil || !encryptedValue.IsBool() {
		return Message{}, invalid("encrypted must be a boolean.", &id)
	}
	encrypted := encryptedValue.Bool()

	hashValue := root.Get("hash")
	if hashValue == nil || !hashValue.IsString() || utf8Len(hashValue.Str()) > MaxHashBytes || hashValue.Str() == "" {
		return Message{}, invalid("hash is required.", &id)
	}
	hash := hashValue.Str()

	clip := &ClientClip{ID: id, Payload: payload, Encrypted: encrypted, Hash: hash}
	if ValidateClip(clip, cfg) {
		return Message{Kind: KindClip, Clip: clip}, nil
	}
	return Message{}, invalid("clip validation failed.", &id)
}

func parsePong(root *Node) (Message, *Error) {
	clientTime, ok := ParseFlexibleTime(root, "clientTimeUtc")
	if !ok {
		return Message{}, invalid("clientTimeUtc must be UTC RFC3339.", nil)
	}

	return Message{Kind: KindPong, Pong: &ClientPong{ClientTimeUtc: clientTime}}, nil
}

// ValidateHello 对应 C# ValidateHello。
func ValidateHello(hello *ClientHello, cfg *config.RuntimeConfig) bool {
	return utf8Len(hello.ClientID) > 0 && utf8Len(hello.ClientID) <= MaxNameBytes &&
		utf8Len(hello.ClientName) <= MaxNameBytes &&
		(hello.Snapshot == nil || validateSnapshot(hello.Snapshot, cfg))
}

func validateSnapshot(snapshot *ClipSnapshot, cfg *config.RuntimeConfig) bool {
	return len(snapshot.Payload) > 0 &&
		utf8Len(snapshot.Payload) <= cfg.Limits.MaxTextBytes &&
		utf8Len(snapshot.Hash) <= MaxHashBytes &&
		isUTCZeroOffset(snapshot.LocalModifiedAtUtc)
}

// ValidateClip 对应 C# ValidateClipMessage。
func ValidateClip(message *ClientClip, cfg *config.RuntimeConfig) bool {
	return len(message.ID) > 0 &&
		utf8Len(message.ID) <= MaxIdBytes &&
		len(message.Payload) > 0 &&
		utf8Len(message.Payload) <= cfg.Limits.MaxTextBytes &&
		len(message.Hash) > 0 &&
		utf8Len(message.Hash) <= MaxHashBytes
}

// CheckFrameSize 对应 C# CheckFrameSize。
func CheckFrameSize(frameLength int, cfg *config.RuntimeConfig) bool {
	return frameLength > 0 && frameLength <= cfg.Limits.MaxFrameBytes
}

// CheckPayloadSize 对应 C# CheckPayloadSize。
func CheckPayloadSize(payload string, cfg *config.RuntimeConfig) bool {
	return utf8Len(payload) <= cfg.Limits.MaxTextBytes
}

func getReferenceID(root *Node) *string {
	if root.IsObject() {
		if id := root.Get("id"); id != nil && id.IsString() {
			value := id.Str()
			return &value
		}
	}
	return nil
}

// hasUniqueKnownProperties 复刻 C# HasUniqueKnownProperties：按文档序，
// 第一个"不在 known 集合中或重复"的字段名即失败原因。
func hasUniqueKnownProperties(root *Node, known []string) (string, bool) {
	knownSet := make(map[string]struct{}, len(known))
	for _, key := range known {
		knownSet[key] = struct{}{}
	}
	seen := make(map[string]struct{}, len(known))
	for _, member := range root.Members() {
		_, knownMatch := knownSet[member.Key]
		_, dup := seen[member.Key]
		if !knownMatch || dup {
			return member.Key, false
		}
		seen[member.Key] = struct{}{}
	}
	return "", true
}

// tryGetUint64 复刻 C# TryGetUInt64：数字形态检查（无 -/./e/E）+ ParseUint。
func tryGetUint64(root *Node, name string) (uint64, bool) {
	element := root.Get(name)
	if element == nil || !element.IsNumber() {
		return 0, false
	}

	raw := element.RawNumber()
	if strings.HasPrefix(raw, "-") || strings.ContainsAny(raw, ".eE") {
		return 0, false
	}

	value, err := strconv.ParseUint(raw, 10, 64)
	if err != nil {
		return 0, false
	}
	return value, true
}

// ParseFlexibleTime 复刻 C# TryGetUtcDateTime：必须以 'Z' 结尾；
// 接受秒级（yyyy-MM-ddTHH:mm:ssZ）或 ISO 往返（含 1-7 位小数秒）两种形态；
// 偏移必须为零（Z 即零偏移）。
func ParseFlexibleTime(root *Node, name string) (time.Time, bool) {
	element := root.Get(name)
	if element == nil || !element.IsString() {
		return time.Time{}, false
	}

	text := element.Str()
	if !strings.HasSuffix(text, "Z") {
		return time.Time{}, false
	}

	value, ok := parseFlexibleUTC(text)
	if !ok {
		return time.Time{}, false
	}
	return value, true
}

func parseFlexibleUTC(text string) (time.Time, bool) {
	body := strings.TrimSuffix(text, "Z")
	const dateLayout = "2006-01-02"
	const secondsLayout = "15:04:05"
	tIndex := strings.IndexByte(body, 'T')
	if tIndex < 0 {
		return time.Time{}, false
	}
	datePart, timePart := body[:tIndex], body[tIndex+1:]

	day, err := time.Parse(dateLayout, datePart)
	if err != nil {
		return time.Time{}, false
	}

	fractionDigits := 0
	if dot := strings.IndexByte(timePart, '.'); dot >= 0 {
		fractionDigits = len(timePart) - dot - 1
		if fractionDigits < 1 || fractionDigits > 7 {
			return time.Time{}, false
		}
	}

	clock, err := time.Parse(secondsLayout, timePart)
	if err != nil {
		return time.Time{}, false
	}

	result := time.Date(
		day.Year(), day.Month(), day.Day(),
		clock.Hour(), clock.Minute(), clock.Second(),
		fractionNanoseconds(timePart, fractionDigits),
		time.UTC)
	return result, true
}

func fractionNanoseconds(timePart string, digits int) int {
	if digits == 0 {
		return 0
	}
	dot := strings.IndexByte(timePart, '.')
	frac := timePart[dot+1 : dot+1+digits]
	nanos, _ := strconv.Atoi(frac)
	for i := digits; i < 9; i++ {
		nanos *= 10
	}
	return nanos
}

func isUTCZeroOffset(t time.Time) bool {
	_, offset := t.Zone()
	return offset == 0
}

func utf8Len(s string) int {
	return len(s)
}
